packages/serverpod/lib/src/server/experimental_features.dart

```diff
@@ -7,6 +9,85 @@ class ExperimentalFeatures {
   /// List of [DiagnosticEventHandler] that will be called for all diagnostic events.
   final List<DiagnosticEventHandler>? diagnosticEventHandlers;
 
+  /// List of [Middleware] to process HTTP requests.
+  ///
+  /// Middleware is executed in the order it appears in this list using
+  /// relic's [Pipeline]. Each middleware can intercept, modify, or
+  /// short-circuit requests before they reach the core handler.
+  ///
+  /// **Note**: This is an experimental feature and the API may change in
+  /// future versions. Middleware uses relic's native [Middleware] typedef
+  /// which wraps handlers to add functionality.
+  ///
+  /// **WebSocket Bypass**: WebSocket upgrade requests (`/websocket`,
+  /// `/v1/websocket`) bypass middleware entirely as they require direct
+  /// access to the connection upgrade mechanism.
+  ///
+  /// ## Example Usage
+  ///
+  /// ```dart
+  /// // Create middleware before Serverpod instance
+  /// final metrics = HttpMetrics();
+  /// final middleware = [
+  ///   loggingMiddleware(logger: (msg) => print(msg)),
+  ///   metricsMiddleware(metrics),
+  /// ];
+  ///
+  /// final pod = await Serverpod(
+  ///   args,
+  ///   serializationManager,
+  ///   endpoints,
+  ///   experimentalFeatures: ExperimentalFeatures(
+  ///     middleware: middleware,
+  ///   ),
+  /// );
+  /// ```
+  ///
+  /// ## Creating Custom Middleware
+  ///
+  /// ### Using createMiddleware helper
+  /// ```dart
+  /// final myMiddleware = createMiddleware(
+  ///   onRequest: (request) {
+  ///     print('Request: ${request.method} ${request.requestedUri}');
+  ///     return null; // Continue to next handler
+  ///   },
+  ///   onResponse: (response) {
+  ///     print('Response: ${response.statusCode}');
+  ///     return response;
+  ///   },
```

Comment:
I suggest we don't put `createMiddleware` in the limelight. This is inherited from shelf, and I prefer if people don't use it and become acquainted with the canonical way